File: apps/a-demo-be-java-vertx/src/test/java/com/ademobejavx/unit/steps/UnitUserAccountSteps.java
package com.ademobejavx.unit.steps;

import com.ademobejavx.unit.UnitFactory;
import com.ademobejavx.support.DirectCallService;
import com.ademobejavx.support.ScenarioState;
import com.ademobejavx.support.ServiceResponse;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;

public class UnitUserAccountSteps {

    private final ScenarioState state;

    public UnitUserAccountSteps(ScenarioState state) {
        this.state = state;
    }

    private DirectCallService svc() {
        return UnitFactory.getService();
    }

    @When("^alice sends GET /api/v1/users/me$")
    public void aliceSendsGetMe() throws Exception {
        String token = state.getAccessToken();
        ServiceResponse response = svc().getMe(token);
        state.setLastResponse(response);
    }

    @When("^alice sends PATCH /api/v1/users/me with body [{] \"displayName\": \"([^\"]*)\" [}]$")
    public void aliceSendsPatchMe(String displayName) throws Exception {
        String token = state.getAccessToken();
        ServiceResponse response = svc().updateMe(token, displayName);
        state.setLastResponse(response);
    }

    @When("^alice sends POST /api/v1/users/me/password with body [{] \"oldPassword\": \"([^\"]*)\", \"newPassword\": \"([^\"]*)\" [}]$")
    public void aliceSendsChangePassword(String oldPassword, String newPassword) throws Exception {
        String token = state.getAccessToken();
        ServiceResponse response = svc().changePassword(token, oldPassword, newPassword);
        state.setLastResponse(response);
    }

    @When("^alice sends POST /api/v1/users/me/deactivate$")
    public void aliceSendsDeactivate() throws Exception {
        String token = state.getAccessToken();
        ServiceResponse response = svc().deactivateMe(token);
        state.setLastResponse(response);
    }

    @Given("^alice has deactivated her own account via POST /api/v1/users/me/deactivate$")
    public void aliceHasDeactivatedHerAccount() throws Exception {
        aliceSendsDeactivate();
    }

    @When("^the client sends GET /api/v1/users/me with alice's access token$")
    public void clientSendsGetMeWithAlicesToken() throws Exception {
        String token = state.getAccessToken();
        ServiceResponse response = svc().getMe(token);
        state.setLastResponse(response);
    }
}

File: apps/demo-be-kotlin-ktor/src/test/kotlin/com/organiclever/demoktkt/unit/steps/UnitUserAccountSteps.kt
package com.organiclever.demoktkt.unit.steps

import io.cucumber.java.en.When

class UnitUserAccountSteps {

  @When("alice sends GET \\/api\\/v1\\/users\\/me")
  fun aliceSendsGetUsersMe() {
    val token = UnitTestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val (status, body) = UnitHttpHelper.get("/api/v1/users/me", token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @When("^alice sends PATCH /api/v1/users/me with body \\{ \"display_name\": \"([^\"]+)\" \\}$")
  fun aliceSendsPatchUsersMeWithBody(displayName: String) {
    val token = UnitTestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val body = """{"display_name":"$displayName"}"""
    val (status, respBody) = UnitHttpHelper.patch("/api/v1/users/me", body, token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = respBody
  }

  @When(
    "^alice sends POST /api/v1/users/me/password with body \\{ \"old_password\": \"([^\"]+)\", \"new_password\": \"([^\"]+)\" \\}$"
  )
  fun aliceSendsPostChangePasswordWithBody(oldPassword: String, newPassword: String) {
    val token = UnitTestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val body = """{"old_password":"$oldPassword","new_password":"$newPassword"}"""
    val (status, respBody) = UnitHttpHelper.post("/api/v1/users/me/password", body, token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = respBody
  }

  @When("alice sends POST \\/api\\/v1\\/users\\/me\\/deactivate")
  fun aliceSendsPostDeactivate() {
    val token = UnitTestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val (status, respBody) = UnitHttpHelper.post("/api/v1/users/me/deactivate", "", token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = respBody
  }
}
